hibernate-core/src/main/java/org/hibernate/boot/SessionFactoryBuilder.java

```diff
@@ -128,16 +129,16 @@
 	SessionFactoryBuilder applyInterceptor(Interceptor interceptor);
 
 	/**
-	 * Names an interceptor Class to be applied to the SessionFactory, which in turn means it will be used by all
+	 * Names an interceptor Supplier to be applied to the SessionFactory, which in turn means it will be used by all
 	 * Sessions unless one is explicitly specified in {@link org.hibernate.SessionBuilder#interceptor}
 	 *
-	 * @param statelessInterceptorClass The interceptor class
+	 * @param statelessInterceptorSupplier The interceptor supplier
 	 *
 	 * @return {@code this}, for method chaining
 	 *
 	 * @see org.hibernate.cfg.AvailableSettings#SESSION_SCOPED_INTERCEPTOR
 	 */
-	SessionFactoryBuilder applyStatelessInterceptor(Class<? extends Interceptor> statelessInterceptorClass);
+	SessionFactoryBuilder applyStatelessInterceptor(Supplier<? extends Interceptor> statelessInterceptorSupplier);
```

Comment:
We should not modify the existing API. We could add a new method, preferably a default one.

Reply:
Would you have a `SessionFactoryBuilder applyStatelessInterceptor(Supplier<? extends Interceptor> statelessInterceptorSupplier);` on the interface above too?

Reply:
I'm not sure how we could have the `SessionFactoryOptions` as it was and incorporate the `Supplier` as its old signature is

`Class<? extends Interceptor> getStatelessInterceptorImplementor();`

Meaning we'd have to convert from supplier to` Class<? extends Interceptor>`

Reply:
For `applyStatelessInterceptor`, we can have 2 overloaded methods. The same for the get part:

- `getStatelessInterceptorImplementor`
- `getStatelessInterceptorSupplier`

So, when building an Interceptor, we check both. If the user supplies both, we should raise an Exception. Otherwise, we use whatever was provided.

This way, we extend the current functionality by adding the new feature instead of modifying the existing interface API or the existing logic that is already in use.

Reply:
I'll have to add a `getStatelessInterceptorSupplier` method to the `SessionFactoryOptionsState` and `SessionFactoryOptions` interfaces otherwise it wont work. Either way this will still break backwards compatibility as upgrading users will have to implement the new supplier methods.

Reply:
Not necessarily. We could add a default method to the interfaces that simply returns a lazy Supplier that delegates the call to `getStatelessInterceptorImplementor`.
Then, in the `SessionFactoryBuilderImpl` we overwrite the default method and add your logic.

Reply:
@vladmihalcea I've pushed the changes you reccommended